draft-ietf-tls-tls13.md

```diff
@@ -2917,28 +2906,49 @@ certificate_authorities
   root CA or for a subordinate CA; thus, this message can be used to
   describe known roots as well as a desired authorization space.  If
   the certificate_authorities list is empty, then the client MAY
-  send any certificate of the appropriate ClientCertificateType,
-  unless there is some external arrangement to the contrary.
+  send any certificate that meets the rest of the selection criteria 
+  in the CertificateRequest, unless there is some external arrangement 
+  to the contrary.
+
+certificate_extensions
+: A list of certificate extension OIDs [RFC5280] with their allowed
+  values, represented in DER-encoded format. Some certificate
+  extension OIDs allow multiple values (e.g. Extended Key Usage).
+  If the server has included a non-empty certificate_extensions list,
+  the client certificate MUST contain all of the specified extension
+  OIDs that the client recognizes. For each extension OID recognized
+  by the client, all of the specified values MUST be present in the
+  client certificate (but the certificate MAY have other values as 
+  well). However, the client MUST ignore and skip any unrecognized 
+  certificate extension OIDs. If the client has ignored some of the 
+  required certificate extension OIDs, and supplied a certificate 
+  that does not satisfy the request, the server MAY at its discretion 
+  either continue the session without client authentication, or 
+  terminate the session with a fatal unsupported_certificate alert.
+
+  PKIX RFCs define a variety of certificate extension OIDs and their 
+  corresponding value types. Depending on the type, matching 
+  certificate extension values are not necessarily bitwise-equal. It 
+  is expected that TLS implementations will rely on their PKI 
+  libraries to perform certificate selection using certificate 
+  extension OIDs.
+
+  This document defines matching rules for two standard certificate 
+  extensions defined in [RFC5280]:
+
+  - Key Usage extension in a certificate matches the request when all 
```

Comment:
Totally.

From: ekr [mailto:notifications@github.com]
Sent: Wednesday, October 14, 2015 2:10 PM
To: tlswg/tls13-spec <tls13-spec@noreply.github.com>
Cc: Andrei Popov <Andrei.Popov@microsoft.com>
Subject: Re: [tls13-spec] Removing certificate_types and adding certificate_extensions (#290)


In draft-ietf-tls-tls13.md<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2ftlswg%2ftls13-spec%2fpull%2f290%23discussion_r42053588&data=01%7c01%7cAndrei.Popov%40microsoft.com%7cb1232f8f6950472d81ae08d2d4dbcd1e%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=yQukDgyxqtC0QNoD6cSrYvbeRnQzRilfsh5le0RcDbg%3d>:

> +  required certificate extension OIDs, and supplied a certificate

> +  that does not satisfy the request, the server MAY at its discretion

> +  either continue the session without client authentication, or

> +  terminate the session with a fatal unsupported_certificate alert.

> +

> +  PKIX RFCs define a variety of certificate extension OIDs and their

> +  corresponding value types. Depending on the type, matching

> +  certificate extension values are not necessarily bitwise-equal. It

> +  is expected that TLS implementations will rely on their PKI

> +  libraries to perform certificate selection using certificate

> +  extension OIDs.

> +

> +  This document defines matching rules for two standard certificate

> +  extensions defined in [RFC5280]:

> +

> +  - Key Usage extension in a certificate matches the request when all

Nit: I would put a "The" in front of this sentence and the one below.

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2ftlswg%2ftls13-spec%2fpull%2f290%2ffiles%23r42053588&data=01%7c01%7cAndrei.Popov%40microsoft.com%7cb1232f8f6950472d81ae08d2d4dbcd1e%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=56gA8nxBjbj7IGZIIczSwRaRsdOErCgowL4kVZfS6CE%3d>.